simtools/db/db_array_elements.py

```diff
@@ -0,0 +1,122 @@
+"""Retrieval of array elements from the database."""
+
+from functools import cache
+
+from pymongo import ASCENDING
+
+
+@cache
+def get_array_elements(db, model_version, collection):
+    """
+    Get all array element names and their design model for a given DB collection.
+
+    Uses the 'design_model' parameter to determine the design model of the array element.
+    Assumes that a design model is defined for every array element.
+
+    Parameters
+    ----------
+    db: DBHandler
+        Instance of the database handler
+    model_version: str
+       Model version.
+    collection: str
+        Database collection (e.g., telescopes, calibration_devices)
+
+    Returns
+    -------
+    dict
+        Dict with array element names found and their design model
+
+    Raises
+    ------
+    ValueError
+        If query for collection name not implemented.
+    KeyError
+        If array element entry in the database is incomplete.
+
+    """
+    db_collection = db.get_collection(db_name=None, collection_name=collection)
+
+    query = {"version": db.model_version(model_version)}
+    results = db_collection.find(query, {"instrument": 1, "value": 1, "parameter": 1}).sort(
+        "instrument", ASCENDING
+    )
+
+    _all_available_array_elements = {}
+    for doc in results:
+        try:
+            if doc["parameter"] == "design_model":
+                _all_available_array_elements[doc["instrument"]] = doc["value"]
+        except KeyError as exc:
+            raise KeyError("Incomplete array element entry in the database.") from exc
+
+    if len(_all_available_array_elements) == 0:
+        raise ValueError(f"No array elements found in DB collection {collection}.")
+
+    return _all_available_array_elements
+
+
+def get_array_element_list_for_db_query(array_element_name, db, model_version, collection):
+    """
+    Get array element name and design model for querying the database.
+
+    Return a list of array element names to be used for querying the database for a given array
+    element. This is in most cases the array element itself and its design model.
+
+    Parameters
+    ----------
+    array_element_name: str
+        Name of the array element model (e.g. MSTN-01).
+    db: DBHandler
+        Instance of the database handler
+    model_version: str
+        Model version.
+    collection: str
+        DB collection to get the array elements from (e.g., telescopes, calibration_devices)
+
+    Returns
+    -------
+    list
+        List of array element model names as used in the DB.
+
+    """
+    _available_array_elements = get_array_elements(db, model_version, collection)
+    try:
+        return [array_element_name, _available_array_elements[array_element_name]]
+    except KeyError:
+        pass
+
+    if array_element_name in _available_array_elements.values():
```

Comment:
When would this if statement return true? I don't understand why it would since the get_array_elements function returns a dictionary with the instrument name as key and the design model as the value. Do we have situations where for eg. the MSTN-01 has the design model of MSTN-01 instead of MSTN-design? If so, what's the point of having separate design models? Even in that case MSTN-01 would exist as a key in the dictionary too.

Reply:
This is e.g. the case for `array_element_name="MSTS-design"` - if you want to get the design model then the first statement with `return [array_element_name, _available_array_elements[array_element_name]]` will raise a KeyError.

The corresponding test is [here](https://github.com/gammasim/simtools/blob/9ca0b173df788c142f5532ff6921a98bf93faccb/tests/unit_tests/db/test_db_array_elements.py#L59).